keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -156,40 +184,51 @@ type PodAntiAffinity struct {
 }
 ```
 
-And it's only effective when (1) it's not nil and (2) "hard" affinity
-requirements (i.e.  `RequiredDuringSchedulingIgnoredDuringExecution`) are
-defined)
+### Design B
+
+Unlike Design A, `EvenSpreading` acts as a standalone spec and applies to
+`pod.spec`. Similarly it's only effective when it's not nil.
+
+```go
+type PodSpec struct {
+    EvenSpreading *EvenSpreading
+    ......
+}
+```
 
-API of `EvenSpreading` is defined as below:
+Inside `EvenSpreading`, we need hard affinityTerms (similar with
+`PodAffinityTerm`) and soft affinityTerms (similar with
+`WeightedPodAffinityTerm`). This describes when we perform even distribution,
+which pods are considered as a group.
 
 ```go
 type EvenSpreading struct {
     // MaxSkew describes the degree of imbalance of pods spreading.
     // Default value is 1 and 0 is not allowed.
     MaxSkew int32
-    // TopologyKey defines where pods are placed evenly
-    // - for NodeAffinity, it can be a well-known key such as
-    //   "failure-domain.beta.kubernetes.io/region" or a self-defined key
-    // - for PodAffinity and PodAntiAffinity, it defaults to
-    //   "kubernetes.io/hostname" due to performance concern
-    TopologyKey string
+    // Similar with the same field in PodAffinity/PodAntiAffinity
+    // +optional
+    RequiredDuringSchedulingIgnoredDuringExecution []PodAffinityTerm
+    // Similar with the same field in PodAffinity/PodAntiAffinity
+    // +optional
+    PreferredDuringSchedulingIgnoredDuringExecution []WeightedPodAffinityTerm
```

Comment:
IIUC, the "Preferred" version will be the same as our current preferred inter-pod anti-affinity. Existence of this item is only justified if we plan to remove preferred anti-affinity.

Reply:
> the "Preferred" version will be the same as our current preferred inter-pod anti-affinity

It's more similar as podAffinity, for both "required" and "preferred" versions. It's just that additionally we need to implement `MaxSkew`.

The whole Design B works independently with any other predicate/priority.

Reply:
Oh wait, I missed a top level `TopologyKey` along with `MaxSkew`. Without it, we can't identify which specific topology we're gonna place the matching pods evenly. (b/c we support `[]PodAffinityTerms`, and each term can has different topology key)

Reply:
BTW: the top level `TopologyKey` can be renamed to `DistributedBy` (or `DistributionKey`), so as not to cause confusion with `TopologyKey` which works inside each affinityTerm.

Reply:
IMO we should remove PodAffinityTerm and WeightedPodAffinityTerm and add a NodeSelector. NodeSelector allows users to specify where pods should be scheduled and what subset of nodes even spreading is applied. For example, a user should be able to specify that their pods should be scheduled in 3 out of 4 available zones and pods should be spread evenly among those 3 zones.
We don't need to specify pod affinity here at all. It complicates the design and I think the API becomes hard to use. Besides, if pod affinity is needed, users can add inter-pod affinity rules to their pod spec. Instead, we should add a pod selector which is a label selector and a namespace only. Unlike PodAffinityTerm, this one does not have a topology key.

Reply:
@bsalamat If we enforce users to use `NodeSelector`, they have to know the node labels in advance, so as to specify the exact selector. And the yamls are not portable either.

On the other hand, I totally agree current API design is complicated, and not that easy to use, but it's expressive. So generally it's basically a similar debate of NodeAffinity vs. PodAffinity. You can't have all the benefits in one solution.

Another solution is to make a tradeoff in between. Just like you pointed out, we can flatten out `[]PodAffinityTerm` to only have `LabelSelector` and `Namespaces`, and they're in the same level with the global `topologyKey`:

```go
type EvenSpreading struct {
    // MaxSkew describes the degree of imbalance of pods spreading.
    // Default value is 1 and 0 is not allowed.
    MaxSkew int32
    // TopologyKey defines where pods are placed evenly
    TopologyKey string
    // Required defines whether it's a hard requirement or soft (preferred) one
    Required bool
	// A label query over a set of resources, in this case pods.
	// +optional
	LabelSelector *metav1.LabelSelector
	// namespaces specifies which namespaces the labelSelector applies to (matches against);
	// null or empty list means "this pod's namespace"
	// +optional
	Namespaces []string
}
```

WDYT?

Reply:
But don't get me wrong, I'm open to any solution. And using `NodeSelector` really makes implementation easier :)

Reply:
@bsalamat I rethink a little on `NodeSelector`. I'm afraid it's not an option. Because `NodeSelector` can only tell "whether a pod can be placed on a node", it can't tell "which pods should be treated as a group". Without the scope of pods, we're unable to do pods spreading.

Why we thought it might work is we implicitly assume either (1) the pods belong to the same application form a group, or (2) pods has exactly the same `NodeSelector` form a group - but neither is correct.

So IMO we should still stick to "PodSelector".

Reply:
omg, i think i just now understood the presence of two different topology keys, one in the podAffinityTerms and one outside. I think its getting too complicated to use. I think we need to think harder here.

Reply:
@Huang-Wei Sorry if I was not clear. I think the API should specify all the following pieces of information:
1. Which pods should be spread (needs a pod selector).
1. What topology pods should be spread to (e.g., node, zone, region, etc.)
1. What set of nodes should be considered. This is needed to filter out some nodes. For example, topology may be "zone", but a user may want to filter out one of the zones.
1. MaxSkew as described in the KEP.
1. Whether this is preferred or required.

This is what I have in mind:

```go

type PodSelectorTerm struct {
	// A label query over a set of resources, in this case pods.
	// +optional
	LabelSelector *metav1.LabelSelector
	// namespaces specifies which namespaces the labelSelector applies to (matches against);
	// null or empty list means "this pod's namespace"
	// +optional
	Namespaces []string
}

type EvenSpreading struct {
    MaxSkew        int32
    TopologyKey    string
    PodSelector    []PodSelectorTerm
    Required       bool  // vs. preffered
    // +optional
    NodeSelector   v1.NodeSelector
}
```

Reply:
After talking with @Huang-Wei offline, I think we can drop the NodeSelector from `EvenSpreading` and let the predicate process an external `NodeAffinity` if there is one in the pod spec.

Reply:
TODO: I will document it somewhere.

Reply:
@bsalamat do you mean the NodeAffinity specified in `Affinity *Affinity` ?

Reply:
@krmayankk Yes. NodeAffinity and NodeSelector are two other predicates that the scheduler has. This feature will take those into account.